flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/FlinkDistribution.java

```diff
@@ -0,0 +1,219 @@
+		}
+
+		for (AutoCloseable fileToDelete : filesToDelete) {
+			try {
+				fileToDelete.close();
+			} catch (Exception e) {
+				LOG.error("Failure while cleaning up file.", e);
+			}
+		}
+	}
+
+	public void startFlinkCluster() throws IOException {
+		AutoClosableProcess.runBlocking("Start Flink cluster", bin.resolve("start-cluster.sh").toAbsolutePath().toString());
+
+		final OkHttpClient client = new OkHttpClient();
+
+		final Request request = new Request.Builder()
+			.get()
+			.url("http://localhost:8081/taskmanagers")
+			.build();
+
+		Exception reportedException = null;
+		for (int x = 0; x < 30; x++) {
+			try (Response response = client.newCall(request).execute()) {
+				if (response.isSuccessful()) {
+					final String json = response.body().string();
+					final JsonNode taskManagerList = OBJECT_MAPPER.readTree(json)
+						.get("taskmanagers");
+
+					if (taskManagerList != null && taskManagerList.size() > 0) {
+						LOG.info("Dispatcher REST endpoint is up.");
+						return;
+					}
+				}
+			} catch (IOException ioe) {
+				reportedException = ExceptionUtils.firstOrSuppressed(ioe, reportedException);
+			}
+
+			LOG.info("Waiting for dispatcher REST endpoint to come up...");
+			try {
+				Thread.sleep(1000);
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				reportedException = ExceptionUtils.firstOrSuppressed(e, reportedException);
+			}
+		}
+		throw new AssertionError("Dispatcher REST endpoint did not start in time.", reportedException);
+	}
+
+	public void stopFlinkCluster() throws IOException {
+		AutoClosableProcess.runBlocking("Stop Flink Cluster", bin.resolve("stop-cluster.sh").toAbsolutePath().toString());
+	}
+
+	public void copyOptJarsToLib(String jarNamePattern) throws FileNotFoundException, IOException {
+		final Optional<Path> reporterJarOptional = Files.walk(opt)
+			.filter(path -> path.getFileName().toString().startsWith("flink-metrics-prometheus"))
+			.findFirst();
+		if (reporterJarOptional.isPresent()) {
+			final Path optReporterJar = reporterJarOptional.get();
+			final Path libReporterJar = lib.resolve(optReporterJar.getFileName());
+			Files.copy(optReporterJar, libReporterJar);
+			filesToDelete.add(new AutoClosablePath(libReporterJar));
+		} else {
+			throw new FileNotFoundException("No jar could be found matching the pattern " + jarNamePattern + ".");
+		}
+	}
+
+	public void appendConfiguration(Configuration config) throws IOException {
+		final Configuration mergedConfig = new Configuration();
+		mergedConfig.addAll(defaultConfig);
+		mergedConfig.addAll(config);
+
+		final List<String> configurationLines = mergedConfig.toMap().entrySet().stream()
+			.map(entry -> entry.getKey() + ": " + entry.getValue())
+			.collect(Collectors.toList());
+
+		Files.write(conf.resolve("flink-conf.yaml"), configurationLines);
+	}
+
+	public Stream<String> searchAllLogs(Pattern pattern, Function<Matcher, String> matchProcessor) throws IOException {
```

Comment:
Can be a future JIRA, but it could make sense to have a separate class with log parsing utilities, instead of putting it here with the `startCluster`, `stopCluster`...

Reply:
No other class will get access to the log path, and as such this is the only place to put it. All accesses to flink-dist will go through this, without exposing any resources contained within to the outside as this will only cause problems.
Effectively `FlinkDistribution` acts as a monitor to interactions with flink-dist.

Reply:
In the long-term I'd like to get rid of log-access completely.

If you cannot write a test without accessing a log it usually highlights an issue with how information is exposed to the user. IMO logs should be last place to get information, with the client/WebUI/REST API being the first.

In this case for example it shouldn't be necessary for me to iterate over all logs to find out which ports are being used by the reporter, this should be exposed as meta information for each JM/TM by the REST API.

Reply:
I totally agree with you! But it is a pretty big effort to achieve that and also even using REST for now is not that safe, as updates happen asynchronously.

Reply:
updates happening asynchronously isn't an issue in regards to safety but of test duration. There are already tests querying metrics via the REST API.